File: java/src/test/java/org/libpetri/core/SubnetDefTest.java
package org.libpetri.core;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link SubnetDef} — open net fragment + interface declaration.
 *
 * <p>Covers MOD-001 (definition), MOD-003/MOD-005 (port/channel declaration via
 * builder), MOD-006 (validation), and MOD-014 ({@link SubnetDef#fromNet}
 * retrofit factory).
 *
 * <p>Tests covering instantiation, per-instance state isolation, nested
 * prefixes, and rich {@link Instance#bindActions} arrive in the next task
 * (the rename pass) — for now, {@link #instantiate_throwsUntilNextTask}
 * pins the current scaffolding behaviour.
 */
class SubnetDefTest {

    record Item(String tag) {}

    // ============================================================
    //  MOD-001: definition + accessors
    // ============================================================

    @Test
    void builder_producesSubnetDef_withName_body_iface() {
        Place<Item> input = Place.of("incoming", Item.class);
        Place<Item> output = Place.of("outgoing", Item.class);

        Transition t = Transition.builder("forward")
            .inputs(Arc.In.one(input))
            .outputs(Arc.Out.and(output))
            .build();

        var def = SubnetDef.builder("Forwarder")
            .transition(t)
            .inputPort("in", input)
            .outputPort("out", output)
            .build();

        assertEquals("Forwarder", def.name());
        assertNotNull(def.body());
        assertNotNull(def.iface());
        assertEquals(Void.class, def.paramType());
        assertTrue(def.body().places().contains(input));
        assertTrue(def.body().places().contains(output));
        assertEquals(1, def.body().transitions().size());
    }

    // ============================================================
    //  MOD-003: ports declared via builder end up on the Interface
    // ============================================================

    @Test
    void inputOutputInoutPort_appearOnInterface() {
        Place<Item> in = Place.of("in_p", Item.class);
        Place<Item> out = Place.of("out_p", Item.class);
        Place<Item> io = Place.of("io_p", Item.class);

        // A single transition that references all three places ensures they
        // are added to the body via auto-collect on the underlying
        // PetriNet.Builder; SubnetDef.Builder requires port places to be
        // present in the body per MOD-006.
        Transition wire = Transition.builder("wire")
            .inputs(Arc.In.one(in))
            .read(io)
            .outputs(Arc.Out.and(out))
            .build();

        var def = SubnetDef.builder("Triple")
            .transition(wire)
            .inputPort("in", in)
            .outputPort("out", out)
            .inoutPort("io", io)
            .build();

        var iface = def.iface();
        assertEquals(3, iface.ports().size());

        assertInstanceOf(Interface.Port.Input.class, iface.port("in").orElseThrow());
        assertInstanceOf(Interface.Port.Output.class, iface.port("out").orElseThrow());
        assertInstanceOf(Interface.Port.InOut.class, iface.port("io").orElseThrow());

        assertEquals(in, iface.port("in").orElseThrow().place());
        assertEquals(out, iface.port("out").orElseThrow().place());
        assertEquals(io, iface.port("io").orElseThrow().place());
    }

    // ============================================================
    //  MOD-005: channels declared via builder end up on the Interface
    // ============================================================

    @Test
    void channel_declared_appearsOnInterface() {
        Place<Item> p = Place.of("p", Item.class);
        Transition attempt = Transition.builder("attempt").read(p).build();

        var def = SubnetDef.builder("RetryPolicy")
            .transition(attempt)
            .channel("attempt", attempt)
            .build();

        var ch = def.iface().channel("attempt").orElseThrow();
        assertEquals(attempt, ch.transition());
        assertInstanceOf(Interface.Channel.SyncChannel.class, ch);
    }

    // ============================================================
    //  MOD-006: validation rejections
    // ============================================================

    @Test
    void build_rejectsPortPlace_notInBody() {
        Place<Item> bodyPlace = Place.of("inBody", Item.class);
        Place<Item> stranger = Place.of("stranger", Item.class);

        // Body references only bodyPlace via a read-arc transition.
        Transition t = Transition.builder("op").read(bodyPlace).build();

        var ex = assertThrows(IllegalArgumentException.class, () ->
            SubnetDef.builder("Bad")
                .transition(t)
                .inputPort("ghost", stranger) // stranger never appears in any arc
                .build());

        assertTrue(ex.getMessage().contains("ghost") || ex.getMessage().contains("stranger"),
            "Error must name the offending port or place: " + ex.getMessage());
    }

    @Test
    void build_rejectsChannelTransition_notInBody() {
        Place<Item> p = Place.of("p", Item.class);
        Transition inBody = Transition.builder("real").read(p).build();
        Transition outOfBody = Transition.builder("ghost").read(p).build();

        var ex = assertThrows(IllegalArgumentException.class, () ->
            SubnetDef.builder("Bad")
                .transition(inBody)
                .channel("ghost", outOfBody)
                .build());

        assertTrue(ex.getMessage().contains("ghost"),
            "Error must name offending channel: " + ex.getMessage());
    }

    @Test
    void build_rejectsDuplicatePortName_withinNamespace() {
        Place<Item> p1 = Place.of("p1", Item.class);
        Place<Item> p2 = Place.of("p2", Item.class);

        var ex = assertThrows(IllegalArgumentException.class, () ->
            SubnetDef.builder("Bad")
                .inputPort("dup", p1)
                .outputPort("dup", p2)
                .build());

        assertTrue(ex.getMessage().contains("dup"),
            "Error must name duplicate port: " + ex.getMessage());
    }

    @Test
    void build_rejectsDuplicateChannelName_withinNamespace() {
        Place<Item> p = Place.of("p", Item.class);
        Transition t1 = Transition.builder("t1").read(p).build();
        Transition t2 = Transition.builder("t2").read(p).build();

        var ex = assertThrows(IllegalArgumentException.class, () ->
            SubnetDef.builder("Bad")
                .transitions(t1, t2)
                .channel("ch", t1)
                .channel("ch", t2)
                .build());

        assertTrue(ex.getMessage().contains("ch"),
            "Error must name duplicate channel: " + ex.getMessage());
    }

    // ============================================================
    //  MOD-014: fromNet retrofit factory
    // ============================================================

    @Test
    void fromNet_producesSubnetDef_whoseBodyEqualsInputNet() {
        Place<Item> a = Place.of("a", Item.class);
        Place<Item> b = Place.of("b", Item.class);
        Transition t = Transition.builder("t")
            .inputs(Arc.In.one(a))
            .outputs(Arc.Out.and(b))
            .build();

        var net = PetriNet.builder("Original").transition(t).build();

        var iface = Interface.builder()
            .inputPort("in", a)
            .build();

        SubnetDef<Void> def = SubnetDef.fromNet(net, iface);

        assertSame(net, def.body(), "fromNet must wrap the supplied net by reference");
        assertEquals(iface, def.iface());
        assertEquals(Void.class, def.paramType());
        assertEquals("Original", def.name());
    }

    // ============================================================
    //  Smoke test: instantiate produces a real Instance now.
    //  (Rich coverage lives in InstanceTest.)
    // ============================================================

    @Test
    void instantiate_producesInstance_withRenamedBody() {
        Place<Item> p = Place.of("p", Item.class);
        Transition t = Transition.builder("t").read(p).build();

        var def = SubnetDef.builder("S").transition(t).build();
        var inst = def.instantiate("pre");

        assertEquals("pre", inst.prefix());
        assertSame(def, inst.def());
        assertNotNull(inst.renamedBody());
        // The renamed body should contain a single, prefixed transition.
        assertEquals(1, inst.renamedBody().transitions().size());
        assertEquals("pre/t", inst.renamedBody().transitions().iterator().next().name());
    }

    // ============================================================
    //  Subnet sealed sum-type membership (MOD-002)
    // ============================================================

    @Test
    void subnetDef_isSubnetOpen_andClosedWrapsPetriNet() {
        Place<Item> p = Place.of("p", Item.class);
        Transition t = Transition.builder("t").read(p).build();
        var def = SubnetDef.builder("S").transition(t).build();

        Subnet asSubnet = def;
        assertInstanceOf(Subnet.Open.class, asSubnet);

        Subnet closed = new Subnet.Closed(PetriNet.builder("X").build());
        assertInstanceOf(Subnet.Closed.class, closed);

        // Pattern-match exhaustiveness — compile error if Subnet gains a new variant.
        String tag = switch (asSubnet) {
            case Subnet.Closed c -> "closed:" + c.net().name();
            case Subnet.Open o -> "open";
        };
        assertEquals("open", tag);
    }
}

File: typescript/tests/core/subnet-def.test.ts
import { describe, it, expect } from 'vitest';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place } from '../../src/core/place.js';
import { one } from '../../src/core/in.js';
import { outPlace } from '../../src/core/out.js';
import { SubnetDef } from '../../src/core/subnet-def.js';
import { Interface } from '../../src/core/interface.js';
import type { Subnet } from '../../src/core/subnet.js';
import { closedSubnet, openSubnet } from '../../src/core/subnet.js';

interface Item {
  readonly tag: string;
}

describe('SubnetDef', () => {
  // ============================================================
  //  MOD-001: definition + accessors
  // ============================================================

  it('builder produces SubnetDef with name, body, iface', () => {
    const input = place<Item>('incoming');
    const output = place<Item>('outgoing');

    const t = Transition.builder('forward')
      .inputs(one(input))
      .outputs(outPlace(output))
      .build();

    const def = SubnetDef.builder('Forwarder')
      .transition(t)
      .inputPort('in', input)
      .outputPort('out', output)
      .build();

    expect(def.name).toBe('Forwarder');
    expect(def.body).toBeDefined();
    expect(def.iface).toBeDefined();
    expect(def.body.places.has(input)).toBe(true);
    expect(def.body.places.has(output)).toBe(true);
    expect(def.body.transitions.size).toBe(1);
  });

  // ============================================================
  //  MOD-003: ports declared via builder end up on the Interface
  // ============================================================

  it('input/output/inout ports appear on Interface', () => {
    const inP = place<Item>('in_p');
    const outP = place<Item>('out_p');
    const ioP = place<Item>('io_p');

    // Single transition references all three places so they're auto-collected
    // into the body via PetriNetBuilder.transition.
    const wire = Transition.builder('wire')
      .inputs(one(inP))
      .read(ioP)
      .outputs(outPlace(outP))
      .build();

    const def = SubnetDef.builder('Triple')
      .transition(wire)
      .inputPort('in', inP)
      .outputPort('out', outP)
      .inoutPort('io', ioP)
      .build();

    const iface = def.iface;
    expect(iface.ports.size).toBe(3);

    expect(iface.port('in')?.direction).toBe('input');
    expect(iface.port('out')?.direction).toBe('output');
    expect(iface.port('io')?.direction).toBe('inout');

    expect(iface.port('in')?.place).toBe(inP);
    expect(iface.port('out')?.place).toBe(outP);
    expect(iface.port('io')?.place).toBe(ioP);
  });

  // ============================================================
  //  MOD-005: channels declared via builder end up on the Interface
  // ============================================================

  it('channel declared appears on Interface', () => {
    const p = place<Item>('p');
    const attempt = Transition.builder('attempt').read(p).build();

    const def = SubnetDef.builder('RetryPolicy')
      .transition(attempt)
      .channel('attempt', attempt)
      .build();

    const ch = def.iface.channel('attempt');
    expect(ch).toBeDefined();
    expect(ch?.transition).toBe(attempt);
  });

  // ============================================================
  //  MOD-006: validation rejections
  // ============================================================

  it('build rejects port place not in body', () => {
    const bodyPlace = place<Item>('inBody');
    const stranger = place<Item>('stranger');

    const t = Transition.builder('op').read(bodyPlace).build();

    expect(() =>
      SubnetDef.builder('Bad')
        .transition(t)
        .inputPort('ghost', stranger)
        .build(),
    ).toThrow(/ghost|stranger/);
  });

  it('build rejects channel transition not in body', () => {
    const p = place<Item>('p');
    const inBody = Transition.builder('real').read(p).build();
    const outOfBody = Transition.builder('ghost').read(p).build();

    expect(() =>
      SubnetDef.builder('Bad')
        .transition(inBody)
        .channel('ghost', outOfBody)
        .build(),
    ).toThrow(/ghost/);
  });

  it('build rejects duplicate port name within port namespace', () => {
    const p1 = place<Item>('p1');
    const p2 = place<Item>('p2');

    expect(() =>
      SubnetDef.builder('Bad')
        .inputPort('dup', p1)
        .outputPort('dup', p2)
        .build(),
    ).toThrow(/dup/);
  });

  it('build rejects duplicate channel name within channel namespace', () => {
    const p = place<Item>('p');
    const t1 = Transition.builder('t1').read(p).build();
    const t2 = Transition.builder('t2').read(p).build();

    expect(() =>
      SubnetDef.builder('Bad')
        .transitions(t1, t2)
        .channel('ch', t1)
        .channel('ch', t2)
        .build(),
    ).toThrow(/ch/);
  });

  // ============================================================
  //  MOD-014: fromNet retrofit factory
  // ============================================================

  it('fromNet produces SubnetDef whose body equals input net', () => {
    const a = place<Item>('a');
    const b = place<Item>('b');
    const t = Transition.builder('t')
      .inputs(one(a))
      .outputs(outPlace(b))
      .build();

    const net = PetriNet.builder('Original').transition(t).build();

    const iface = Interface.builder()
      .inputPort('in', a)
      .build();

    const def = SubnetDef.fromNet(net, iface);

    expect(def.body).toBe(net);
    expect(def.iface).toBe(iface);
    expect(def.name).toBe('Original');
  });

  it('fromNet validates port place membership (MOD-014)', () => {
    const a = place<Item>('a');
    const stranger = place<Item>('stranger');
    const t = Transition.builder('t').read(a).build();
    const net = PetriNet.builder('N').transition(t).build();

    const iface = Interface.builder()
      .inputPort('ghost', stranger)
      .build();

    expect(() => SubnetDef.fromNet(net, iface)).toThrow(/ghost|stranger/);
  });

  it('fromNet validates channel transition membership (MOD-014)', () => {
    const a = place<Item>('a');
    const inNet = Transition.builder('inNet').read(a).build();
    const outOfNet = Transition.builder('outOfNet').read(a).build();
    const net = PetriNet.builder('N').transition(inNet).build();

    const iface = Interface.builder()
      .channel('orphan', outOfNet)
      .build();

    expect(() => SubnetDef.fromNet(net, iface)).toThrow(/orphan/);
  });

  // ============================================================
  //  MOD-051: verify(harness) is implemented in task #15
  //  (See tests/verification/subnet-verify.test.ts for full coverage.)
  // ============================================================

  it('verify is callable (real implementation; full coverage in subnet-verify.test.ts)', async () => {
    const p = place<Item>('p');
    const t = Transition.builder('t').read(p).build();
    const def = SubnetDef.builder('S').transition(t).build();

    // No ports, no properties — the harness builds the synthetic net and
    // returns an empty per-property map. Should not throw "not implemented".
    const result = await def.verify({
      params: undefined as never,
      portInputGenerators: {},
      properties: [],
    });
    expect(result.syntheticNet).toBeDefined();
    expect(result.perProperty.size).toBe(0);
    expect(result.allProven()).toBe(true);
    expect(result.anyViolated()).toBe(false);
  });

  // ============================================================
  //  Subnet discriminated-union exhaustiveness (MOD-002)
  // ============================================================

  it('Subnet pattern matching on kind is exhaustive', () => {
    const p = place<Item>('p');
    const t = Transition.builder('t').read(p).build();
    const def = SubnetDef.builder('S').transition(t).build();

    const open: Subnet = openSubnet(def);
    const closed: Subnet = closedSubnet(PetriNet.builder('X').build());

    function classify(s: Subnet): string {
      switch (s.kind) {
        case 'closed': return `closed:${s.net.name}`;
        case 'open': return `open:${s.def.name}`;
        default: {
          // Exhaustiveness check — `s` must be `never` here.
          const _exhaustive: never = s;
          return _exhaustive;
        }
      }
    }

    expect(classify(open)).toBe('open:S');
    expect(classify(closed)).toBe('closed:X');
  });
});
